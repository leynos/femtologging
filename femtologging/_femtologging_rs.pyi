from typing import Any as _Any, Literal, Union

FemtoLevel: _Any
LevelName = Literal["TRACE", "DEBUG", "INFO", "WARN", "WARNING", "ERROR", "CRITICAL"]
LevelArg = Union[LevelName, FemtoLevel]

FemtoLogger: _Any
FemtoHandler: _Any
FemtoStreamHandler: _Any
FemtoFileHandler: _Any
FemtoRotatingFileHandler: _Any
StreamHandlerBuilder: _Any
FileHandlerBuilder: _Any
RotatingFileHandlerBuilder: _Any
ConfigBuilder: _Any
LoggerConfigBuilder: _Any
FormatterBuilder: _Any
LevelFilterBuilder: _Any
NameFilterBuilder: _Any
FilterBuildError: type[Exception]
HandlerConfigError: type[Exception]
HandlerIOError: type[Exception]

hello: _Any
get_logger: _Any
reset_manager_py: _Any
