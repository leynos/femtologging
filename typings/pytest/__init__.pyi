from typing import (
    Any,
    Callable,
    ContextManager,
    Pattern,
    Type,
    TypeVar,
    overload,
    ParamSpec,
)

P = ParamSpec("P")
R = TypeVar("R", bound=BaseException)

@overload
def fixture(func: Callable[P, R]) -> Callable[P, R]: ...
@overload
def fixture(
    *,
    scope: str | None = ...,
    autouse: bool | None = ...,
    params: list[Any] | None = ...,
    ids: list[str] | Callable[[Any], str] | None = ...,
    name: str | None = ...,
) -> Callable[[Callable[P, R]], Callable[P, R]]: ...
def raises(
    exc: Type[R],
    match: str | Pattern[str] | None = ...,
    *,
    msg: str | None = ...,
) -> ContextManager[R]: ...

class mark:
    @staticmethod
    def parametrize(
        argnames: str | list[str],
        argvalues: list[Any] | tuple[Any, ...] | list[tuple[Any, ...]],
        *,
        ids: list[str] | Callable[[Any], str] | None = ...,
        indirect: bool | list[str] = ...,
        **kwargs: Any,
    ) -> Callable[[Callable[..., Any]], Callable[..., Any]]: ...
