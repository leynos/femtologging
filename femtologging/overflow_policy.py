from __future__ import annotations

from enum import Enum


class OverflowPolicy(Enum):
    """Behaviour when a handler queue is full."""

    DROP = "drop"
    BLOCK = "block"
    TIMEOUT = "timeout"

    def __str__(self) -> str:
        return self.value
