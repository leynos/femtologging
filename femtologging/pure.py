from __future__ import annotations


class FemtoLogger:
    """Simplistic Python implementation used when the Rust module is missing."""

    def __init__(self, name: str) -> None:
        self.name = name

    def log(self, level: str, message: str) -> str:
        """Return the formatted log message."""
        return f"{self.name} [{level}] {message}"


def hello() -> str:
    """Return a friendly greeting from Python."""
    return "hello from Python"


__all__ = ["FemtoLogger", "hello"]
