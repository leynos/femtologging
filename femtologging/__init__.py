"""femtologging package."""

from __future__ import annotations

# Import the Rust extension packaged under this module's namespace first
# to keep imports at the top for linters.
from . import _femtologging_rs as rust
from .overflow_policy import OverflowPolicy
from .config import dictConfig
import logging
import sys
from dataclasses import dataclass
from typing import (
    Callable,
    Iterable,
    TextIO,
    cast,
    overload,
)


hello = rust.hello
FemtoLogger = rust.FemtoLogger
get_logger = rust.get_logger
reset_manager = rust.reset_manager_py
FemtoHandler = rust.FemtoHandler
FemtoStreamHandler = rust.FemtoStreamHandler
FemtoFileHandler = rust.FemtoFileHandler
FemtoRotatingFileHandler = rust.FemtoRotatingFileHandler
HandlerOptions = rust.HandlerOptions
ROTATION_VALIDATION_MSG = rust.ROTATION_VALIDATION_MSG
StreamHandlerBuilder = rust.StreamHandlerBuilder
FileHandlerBuilder = rust.FileHandlerBuilder
RotatingFileHandlerBuilder = rust.RotatingFileHandlerBuilder
ConfigBuilder = rust.ConfigBuilder
LoggerConfigBuilder = rust.LoggerConfigBuilder
FormatterBuilder = rust.FormatterBuilder
LevelFilterBuilder = rust.LevelFilterBuilder
NameFilterBuilder = rust.NameFilterBuilder
FilterBuildError = rust.FilterBuildError
HandlerConfigError = rust.HandlerConfigError
HandlerIOError = rust.HandlerIOError
_force_rotating_fresh_failure = getattr(
    rust, "force_rotating_fresh_failure_for_test", None
)
_clear_rotating_fresh_failure = getattr(
    rust, "clear_rotating_fresh_failure_for_test", None
)

if callable(_force_rotating_fresh_failure) and callable(_clear_rotating_fresh_failure):
    _force_rotating_fresh_failure_for_test = cast(
        Callable[[int, str | None], None],
        _force_rotating_fresh_failure,
    )
    _clear_rotating_fresh_failure_for_test = cast(
        Callable[[], None],
        _clear_rotating_fresh_failure,
    )
else:
    # Feature disabled: expose no-ops that fail loudly when invoked.

    def _force_rotating_fresh_failure_for_test(
        count: int, reason: str | None = None
    ) -> None:
        raise RuntimeError(
            "rotating fresh-failure hook requires the extension built with the 'python' feature"
        )

    def _clear_rotating_fresh_failure_for_test() -> None:
        return


@dataclass
class BasicConfig:
    """Configuration parameters for basicConfig()."""

    level: str | int | None = None
    filename: str | None = None
    stream: TextIO | None = None
    force: bool = False
    handlers: Iterable[FemtoHandler] | None = None


@overload
def basicConfig(config: BasicConfig, /) -> None: ...


@overload
def basicConfig(**kwargs: object) -> None: ...


def basicConfig(config: BasicConfig | None = None, /, **kwargs: object) -> None:
    """Configure the root logger using the builder API.

        Parameters mirror ``logging.basicConfig`` but currently only a subset is
        supported. ``config`` may be a :class:`BasicConfig` instance; if provided,
        its values take precedence over individual parameters. ``level`` may be a
        string or numeric value understood by the standard :mod:`logging` module.
        ``filename`` configures a :class:`FemtoFileHandler`; otherwise a
        :class:`FemtoStreamHandler` targeting ``stderr`` is installed. ``stream``
        may be ``sys.stdout`` to redirect output. ``force`` removes any existing
        handlers from the root logger before applying the new configuration.
        ``handlers`` allows attaching pre‑constructed handlers directly.

        Parameters
        ----------
        config : BasicConfig, optional
            Configuration dataclass providing parameters for ``basicConfig``.

        Other Parameters
        ----------------
        level : str or int, optional
            Logging level. Accepts case-insensitive "TRACE", "DEBUG", "INFO", "WARN",
    "WARNING", "ERROR", and "CRITICAL". "WARN" and "WARNING" are equivalent.
        filename : str, optional
            File to write logs to.
        stream : TextIO, optional
            ``sys.stdout`` or ``sys.stderr``.
        force : bool, default False
            Remove any existing handlers before configuring.
        handlers : Iterable[FemtoHandler], optional
            Pre‑constructed handlers to attach.

        Examples
        --------
        Using a dataclass::

            cfg = BasicConfig(level="INFO")
            basicConfig(cfg)

        Using individual parameters::

            basicConfig(level="INFO")

        Notes
        -----
        ``format`` and ``datefmt`` are intentionally unsupported until formatter
        customisation is implemented.
    """
    allowed = {"level", "filename", "stream", "force", "handlers"}
    unknown = set(kwargs) - allowed
    if unknown:
        name = next(iter(unknown))
        raise TypeError(f"basicConfig() got an unexpected keyword argument {name!r}")

    level: str | int | None
    filename: str | None
    stream: TextIO | None
    force: bool
    handlers: Iterable[FemtoHandler] | None
    if config is not None:
        level = (
            config.level
            if config.level is not None
            else cast(str | int | None, kwargs.get("level"))
        )
        filename = (
            config.filename
            if config.filename is not None
            else cast(str | None, kwargs.get("filename"))
        )
        stream = (
            config.stream
            if config.stream is not None
            else cast(TextIO | None, kwargs.get("stream"))
        )
        force = bool(config.force)
        handlers = (
            config.handlers
            if config.handlers is not None
            else cast(Iterable[FemtoHandler] | None, kwargs.get("handlers"))
        )
    else:
        level = cast(str | int | None, kwargs.get("level"))
        filename = cast(str | None, kwargs.get("filename"))
        stream = cast(TextIO | None, kwargs.get("stream"))
        force = bool(kwargs.get("force", False))
        handlers = cast(Iterable[FemtoHandler] | None, kwargs.get("handlers"))

    _validate_basic_config_params(filename, stream, handlers)

    root = get_logger("root")
    if force:
        root.clear_handlers()

    _configure_handlers(root, handlers, filename, stream)

    _set_logger_level(root, level)


def _has_conflicting_handler_params(
    handlers: Iterable[FemtoHandler] | None,
    filename: str | None,
    stream: TextIO | None,
) -> bool:
    """Check if handlers conflict with filename or stream parameters."""
    return handlers is not None and (filename is not None or stream is not None)


def _validate_basic_config_params(
    filename: str | None,
    stream: TextIO | None,
    handlers: Iterable[FemtoHandler] | None,
) -> None:
    """Validate ``basicConfig`` parameters."""
    if filename and stream:
        raise ValueError("Cannot specify both `filename` and `stream`")

    if _has_conflicting_handler_params(handlers, filename, stream):
        msg = "Cannot specify `handlers` with `filename` or `stream`"
        raise ValueError(msg)

    if stream not in (None, sys.stdout, sys.stderr):
        raise ValueError(
            f"stream must be sys.stdout or sys.stderr, got {type(stream)!r}: {stream!r}"
        )


def _configure_handlers(
    root: FemtoLogger,
    handlers: Iterable[FemtoHandler] | None,
    filename: str | None,
    stream: TextIO | None,
) -> None:
    """Attach or build handlers for the root logger."""
    if handlers is not None:
        for h in handlers:
            root.add_handler(h)
    else:
        _build_and_configure_handler(filename, stream)


def _build_and_configure_handler(
    filename: str | None,
    stream: TextIO | None,
) -> None:
    """Build a handler via the builder API and install it."""
    builder = ConfigBuilder()

    handler_id = "basic_config_handler"
    handler = _create_handler_builder(filename, stream)
    builder.with_handler(handler_id, handler)

    logger_cfg = LoggerConfigBuilder().with_handlers([handler_id])
    builder.with_root_logger(logger_cfg)

    builder.build_and_init()


def _create_handler_builder(
    filename: str | None,
    stream: TextIO | None,
) -> FileHandlerBuilder | StreamHandlerBuilder:
    """Create a handler builder for ``basicConfig``."""
    if filename:
        return FileHandlerBuilder(filename)
    if stream is sys.stdout:
        return StreamHandlerBuilder.stdout()
    return StreamHandlerBuilder.stderr()


def _set_logger_level(root: FemtoLogger, level: str | int | None) -> None:
    """Set the root logger level if provided."""
    if level is not None:
        lvl = logging.getLevelName(level) if isinstance(level, int) else level
        root.set_level(lvl)


__all__ = [
    "FemtoHandler",
    "FemtoLogger",
    "get_logger",
    "reset_manager",
    "FemtoStreamHandler",
    "FemtoFileHandler",
    "FemtoRotatingFileHandler",
    "HandlerOptions",
    "ROTATION_VALIDATION_MSG",
    "StreamHandlerBuilder",
    "FileHandlerBuilder",
    "RotatingFileHandlerBuilder",
    "LevelFilterBuilder",
    "NameFilterBuilder",
    "FilterBuildError",
    "ConfigBuilder",
    "LoggerConfigBuilder",
    "FormatterBuilder",
    "HandlerConfigError",
    "HandlerIOError",
    "OverflowPolicy",
    "BasicConfig",
    "basicConfig",
    "dictConfig",
    "hello",
]
