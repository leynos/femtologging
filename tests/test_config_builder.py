import pytest
import pathlib

from pytest_bdd import given, when, then, scenarios, parsers
from syrupy import SnapshotAssertion

from femtologging import (
    ConfigBuilder,
    FormatterBuilder,
    LoggerConfigBuilder,
    RotatingFileHandlerBuilder,
    StreamHandlerBuilder,
    get_logger,
)


@given("a ConfigBuilder", target_fixture="config_builder")
def given_config_builder() -> ConfigBuilder:
    return ConfigBuilder()


@when('I add formatter "fmt" with format "{level} {message}"')
def add_formatter(config_builder: ConfigBuilder) -> None:
    fmt = FormatterBuilder().with_format("{level} {message}")
    config_builder.with_formatter("fmt", fmt)


@when('I add logger "core" with level "INFO"')
def add_logger(config_builder: ConfigBuilder) -> None:
    logger = LoggerConfigBuilder().with_level("INFO")
    config_builder.with_logger("core", logger)


@when(parsers.parse('I add stream handler "{hid}" targeting "{target}"'))
def add_stream_handler(config_builder: ConfigBuilder, hid: str, target: str) -> None:
    handler = (
        StreamHandlerBuilder.stderr()
        if target.lower() == "stderr"
        else StreamHandlerBuilder.stdout()
    )
    config_builder.with_handler(hid, handler)


@when(parsers.parse('I add logger "{name}" with handler "{handler}"'))
def add_logger_with_handler(
    config_builder: ConfigBuilder, name: str, handler: str
) -> None:
    logger = LoggerConfigBuilder().with_handlers([handler])
    config_builder.with_logger(name, logger)


@when(
    parsers.parse('I add logger "{name}" with level "{level}" and handler "{handler}"')
)
def add_logger_with_level_and_handler(
    config_builder: ConfigBuilder, name: str, level: str, handler: str
) -> None:
    logger = LoggerConfigBuilder().with_level(level).with_handlers([handler])
    config_builder.with_logger(name, logger)


@when(parsers.parse('I set root logger with level "{level}"'))
def set_root(config_builder: ConfigBuilder, level: str) -> None:
    root = LoggerConfigBuilder().with_level(level)
    config_builder.with_root_logger(root)


@then("the configuration matches snapshot")
def configuration_matches_snapshot(
    config_builder: ConfigBuilder, snapshot: SnapshotAssertion
) -> None:
    assert config_builder.as_dict() == snapshot
    config_builder.build_and_init()


@when("I set version 2")
def set_version(config_builder: ConfigBuilder) -> None:
    config_builder.with_version(2)


@then("building the configuration fails")
def build_fails(config_builder: ConfigBuilder) -> None:
    with pytest.raises(ValueError):
        config_builder.build_and_init()


@then(parsers.parse('building the configuration fails with error containing "{msg}"'))
def build_fails_with_value_error(config_builder: ConfigBuilder, msg: str) -> None:
    with pytest.raises(ValueError) as excinfo:
        config_builder.build_and_init()
    assert msg in str(excinfo.value)


@then(
    parsers.parse('building the configuration fails with key error containing "{msg}"')
)
def build_fails_with_key_error(config_builder: ConfigBuilder, msg: str) -> None:
    with pytest.raises(KeyError) as excinfo:
        config_builder.build_and_init()
    assert msg in str(excinfo.value)


@then(parsers.parse('loggers "{first}" and "{second}" share handler "{hid}"'))
def loggers_share_handler(first: str, second: str, hid: str) -> None:
    del hid  # parameter required by step signature
    first_logger = get_logger(first)
    second_logger = get_logger(second)
    h1 = first_logger.handler_ptrs_for_test()
    h2 = second_logger.handler_ptrs_for_test()
    assert h1[0] == h2[0], "handlers should be shared"


scenarios("features/config_builder.feature")


def test_duplicate_formatter_overwrites() -> None:
    """Second formatter with same ID should replace the first."""
    builder = ConfigBuilder()
    fmt1 = FormatterBuilder().with_format("one")
    fmt2 = FormatterBuilder().with_format("two")
    builder.with_formatter("fmt", fmt1)
    builder.with_formatter("fmt", fmt2)
    config = builder.as_dict()
    assert config["formatters"]["fmt"]["format"] == "two", (
        "Later formatter should overwrite earlier one"
    )


def test_duplicate_handler_overwrites() -> None:
    """Second handler with same ID should replace the first."""
    builder = ConfigBuilder()
    handler1 = StreamHandlerBuilder.stderr()
    handler2 = StreamHandlerBuilder.stdout()
    builder.with_handler("console", handler1)
    builder.with_handler("console", handler2)
    logger = LoggerConfigBuilder().with_handlers(["console"])
    builder.with_logger("core", logger)
    builder.with_root_logger(LoggerConfigBuilder().with_level("INFO"))
    config = builder.as_dict()
    assert config["handlers"]["console"]["target"] == "stdout", (
        "Later handler should overwrite earlier one",
    )


def test_rotating_handler_supported(tmp_path: pathlib.Path) -> None:
    """ConfigBuilder should accept rotating file handler builders."""
    builder = ConfigBuilder().with_disable_existing_loggers(True)
    log_path = tmp_path / "rotating.log"
    rotating = (
        RotatingFileHandlerBuilder(str(log_path))
        .with_max_bytes(1024)
        .with_backup_count(3)
    )
    builder.with_handler("rot", rotating)
    builder.with_root_logger(LoggerConfigBuilder().with_handlers(["rot"]))

    # Building should succeed and preserve the rotating handler configuration.
    builder.build_and_init()
    config = builder.as_dict()
    assert config["handlers"]["rot"]["path"] == str(log_path)
    assert config["handlers"]["rot"]["max_bytes"] == 1024
    assert config["handlers"]["rot"]["backup_count"] == 3


def test_duplicate_logger_overwrites() -> None:
    """Second logger with same ID should replace the first."""
    builder = ConfigBuilder()
    logger1 = LoggerConfigBuilder().with_level("INFO")
    logger2 = LoggerConfigBuilder().with_level("ERROR")
    builder.with_logger("core", logger1)
    builder.with_logger("core", logger2)
    builder.with_root_logger(LoggerConfigBuilder().with_level("WARNING"))
    config = builder.as_dict()
    assert config["loggers"]["core"]["level"] == "ERROR", (
        "Later logger should overwrite earlier one"
    )


def test_logger_config_builder_optional_fields_set() -> None:
    """Test that optional fields are included when explicitly set."""
    logger = (
        LoggerConfigBuilder()
        .with_level("DEBUG")
        .with_propagate(False)
        .with_filters(["myfilter"])
        .with_handlers(["console", "file"])
    )
    config = logger.as_dict()
    assert config["level"] == "DEBUG", "Level should be included when set"
    assert config["propagate"] is False, "Propagate should be included when set"
    assert config["filters"] == ["myfilter"], "Filters should be included when set"
    assert config["handlers"] == ["console", "file"], (
        "Handlers should be included when set"
    )


def test_logger_config_builder_optional_fields_omitted() -> None:
    """Test that optional fields are omitted when not set."""
    logger = LoggerConfigBuilder().with_level("WARNING")
    config = logger.as_dict()
    assert config["level"] == "WARN", "Level should be normalised to WARN"
    assert "propagate" not in config, "Propagate should be omitted when not set"
    assert "filters" not in config, "Filters should be omitted when not set"
    assert "handlers" not in config, "Handlers should be omitted when not set"


def test_no_root_logger_behavior() -> None:
    """Test that building without a root logger raises ValueError."""
    builder = ConfigBuilder()
    with pytest.raises(ValueError):
        builder.build_and_init()


def test_unknown_handler_id_raises_key_error() -> None:
    """Building with an unknown handler identifier raises KeyError."""
    builder = ConfigBuilder()
    logger = LoggerConfigBuilder().with_handlers(["missing"])
    builder.with_logger("core", logger)
    builder.with_root_logger(LoggerConfigBuilder().with_level("INFO"))
    with pytest.raises(KeyError, match="missing"):
        builder.build_and_init()


def make_info_stderr_builder() -> ConfigBuilder:
    """Create a builder with an INFO root logger and stderr handler."""
    return (
        ConfigBuilder()
        .with_handler("h", StreamHandlerBuilder.stderr())
        .with_root_logger(LoggerConfigBuilder().with_level("INFO"))
    )


def make_builder_with_logger(logger_name: str) -> ConfigBuilder:
    """Create a builder with INFO root logger, stderr handler, and a named logger."""
    return make_info_stderr_builder().with_logger(
        logger_name, LoggerConfigBuilder().with_handlers(["h"])
    )


def test_disable_existing_loggers_clears_unmentioned() -> None:
    """Loggers not present in new config are disabled."""
    builder = make_builder_with_logger("stale")
    builder.build_and_init()

    stale = get_logger("stale")
    assert stale.handler_ptrs_for_test(), "stale logger should have a handler"

    rebuild = (
        ConfigBuilder()
        .with_root_logger(LoggerConfigBuilder().with_level("INFO"))
        .with_disable_existing_loggers(True)
    )
    rebuild.build_and_init()

    stale = get_logger("stale")
    assert stale.handler_ptrs_for_test() == [], "stale logger should be disabled"


@pytest.mark.parametrize(
    "ancestors",
    [
        ["parent"],
        ["grandparent", "grandparent.parent"],
    ],
    ids=["parent", "grandparent"],
)
def test_disable_existing_loggers_keeps_ancestors(ancestors: list[str]) -> None:
    """Ancestor loggers remain active when their descendants are configured."""
    builder = make_info_stderr_builder()
    for name in ancestors:
        builder = builder.with_logger(name, LoggerConfigBuilder().with_handlers(["h"]))
    builder.build_and_init()

    initial_handlers = {
        name: get_logger(name).handler_ptrs_for_test() for name in ancestors
    }

    child_name = f"{ancestors[-1]}.child"
    rebuild = (
        make_info_stderr_builder()
        .with_logger(child_name, LoggerConfigBuilder().with_handlers(["h"]))
        .with_disable_existing_loggers(True)
    )
    rebuild.build_and_init()

    child = get_logger(child_name)
    assert len(child.handler_ptrs_for_test()) == 1, "child should have one handler"
    for name in ancestors:
        assert get_logger(name).handler_ptrs_for_test() == initial_handlers[name], (
            "ancestor logger should retain its handler"
        )


@pytest.mark.parametrize(
    ("first", "second", "expected"),
    [
        ("INFO", "ERROR", "ERROR"),
        ("ERROR", "INFO", "INFO"),
        ("DEBUG", "WARN", "WARN"),
    ],
    ids=["INFO→ERROR", "ERROR→INFO", "DEBUG→WARN"],
)
def test_root_logger_last_assignment_wins(
    first: str, second: str, expected: str
) -> None:
    """Verify last-write-wins semantics when assigning the root logger multiple times."""
    builder = ConfigBuilder()
    builder.with_root_logger(LoggerConfigBuilder().with_level(first))
    builder.with_root_logger(LoggerConfigBuilder().with_level(second))
    config = builder.as_dict()
    assert config["root"]["level"] == expected, (
        f"Last root logger assignment wins: {first}→{second}"
    )
