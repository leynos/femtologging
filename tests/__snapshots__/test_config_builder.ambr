# serializer version: 1
# name: test_attach_handler_to_multiple_loggers
  dict({
    'disable_existing_loggers': False,
    'handlers': dict({
      'console': dict({
        'target': 'stderr',
      }),
    }),
    'loggers': dict({
      'core': dict({
        'handlers': list([
          'console',
        ]),
      }),
      'worker': dict({
        'handlers': list([
          'console',
        ]),
      }),
    }),
    'root': dict({
      'level': 'INFO',
    }),
    'version': 1,
  })
# ---
# name: test_build_simple_configuration
  dict({
    'disable_existing_loggers': False,
    'formatters': dict({
      'fmt': dict({
        'format': '{level} {message}',
      }),
    }),
    'loggers': dict({
      'core': dict({
        'level': 'INFO',
      }),
    }),
    'root': dict({
      'level': 'WARN',
    }),
    'version': 1,
  })
# ---
